bot/cogs/codeblock/parsing.py

```diff
@@ -0,0 +1,226 @@
+
+    return code_blocks
+
+
+def _is_python_code(content: str) -> bool:
+    """Return True if `content` is valid Python consisting of more than just expressions."""
+    log.trace("Checking if content is Python code.")
+    try:
+        # Attempt to parse the message into an AST node.
+        # Invalid Python code will raise a SyntaxError.
+        tree = ast.parse(content)
+    except SyntaxError:
+        log.trace("Code is not valid Python.")
+        return False
+
+    # Multiple lines of single words could be interpreted as expressions.
+    # This check is to avoid all nodes being parsed as expressions.
+    # (e.g. words over multiple lines)
+    if not all(isinstance(node, ast.Expr) for node in tree.body):
+        log.trace("Code is valid python.")
+        return True
+    else:
+        log.trace("Code consists only of expressions.")
+        return False
+
+
+def _is_repl_code(content: str, threshold: int = 3) -> bool:
+    """Return True if `content` has at least `threshold` number of (I)Python REPL-like lines."""
+    log.trace(f"Checking if content is (I)Python REPL code using a threshold of {threshold}.")
+
+    repl_lines = 0
+    patterns = (_RE_PYTHON_REPL, _RE_IPYTHON_REPL)
+
+    for line in content.splitlines():
+        # Check the line against all patterns.
+        for pattern in patterns:
+            if pattern.match(line):
+                repl_lines += 1
+
+                # Once a pattern is matched, only use that pattern for the remaining lines.
+                patterns = (pattern,)
+                break
+
+        if repl_lines == threshold:
+            log.trace("Content is (I)Python REPL code.")
+            return True
+
+    log.trace("Content is not (I)Python REPL code.")
+    return False
+
+
+def is_python_code(content: str) -> bool:
+    """Return True if `content` is valid Python code or (I)Python REPL output."""
+    dedented = textwrap.dedent(content)
+
+    # Parse AST twice in case _fix_indentation ends up breaking code due to its inaccuracies.
+    return (
+        _is_python_code(dedented)
+        or _is_repl_code(dedented)
+        or _is_python_code(_fix_indentation(content))
+    )
+
+
+def parse_bad_language(content: str) -> Optional[BadLanguage]:
+    """
+    Return information about a poorly formatted Python language in code block `content`.
+
+    If the language is not Python, return None.
+    """
+    log.trace("Parsing bad language.")
+
+    match = _RE_LANGUAGE.match(content)
+    if not match:
+        return None
+
+    return BadLanguage(
+        language=match["lang"],
+        has_leading_spaces=match["spaces"] is not None,
+        has_terminal_newline=match["newline"] is not None,
+    )
+
+
+def _get_leading_spaces(content: str) -> int:
+    """Return the number of spaces at the start of the first line in `content`."""
+    current = content[0]
+    leading_spaces = 0
+
+    while current == " ":
+        leading_spaces += 1
+        current = content[leading_spaces]
```

Comment:
This seems a bit confusing to me, would a for loop that'd do something like this be more fitting?

```py
    for character in content:
        if character == " ":
           increment spaces
        else:
            return spaces

Reply:
I may be biased since I wrote it, but I don't think it's confusing. Anyone else feel a for loop is clearer?

Reply:
A for loop seems a *little bit* more idiomatic here, but IMO either approach is fine in this case.

Reply:
It could also be `return len(itertools.takewhile(lambda c: c == " ", content))`, but I'll change it to a for loop.